### ISSUE TITLE
System.Text.Json.JsonSerializer doesn't serialize properties from derived classes

### ISSUE THREAD

**Issue by mauricio-bv** [dotnet/core#4060]
# Issue Title

System.Text.Json.JsonSerializer doesn't serialize properties from derived classes

# General

I am trying to use the System.Text.Json serialization libray, but I can't get it is serialize it as I used to with Newtonson. Properties in derived classes or Interfaces are not serializing. This issue has also been asked and explained in the link below.
[https://stackoverflow.com/questions/59258655/system-text-json-jsonserializer-doesnt-serialize-properties-from-derived-classe/59536948#59536948]

Is there any serialization option I should set for proper serialization?

**Comment by bartonjs** [comment]
The serializer serializes using whatever type information you gave it.  Static typing on serialization prevents accidental data disclosure from derived types that didn't understand they were being serialized.

If you really want polymorphic serialization, you can accomplish it in one of three different ways:

* JsonSerializer.Serialize<object>(value, ...)
* JsonSerializer.Serialize(value, typeof(object), ...);
* JsonSerializer.Serialize(value, value.GetType(), ...);

If you just have code like `JsonSerializer.Serialize(value, ...)` then generic inference binds it as the static type of the `value` parameter.

**Comment by mauricio-bv** [comment]
Hi Bartonjs.
Your suggestion works for the simple example I provided, but in the example below it does not work (just put the whole code in the Program.cs file of a console application and run). The instances of Vehicle do not serialize properly.

`using System.Collections.Generic;
using System.Text.Json;

namespace NetCoreSerializerProblem
{
    public interface IVehicle : IProduct
    {
        ISpecs Specs { get; set; }
    }

    public interface IProduct
    {
        string Name { get; set; }
    }


    public interface ISpecs
    {
        int Doors { get; set; }
    }

    public class Truck : Vehicle
    {
        public Truck(string name) : base(name) { }
    }

    public abstract class Vehicle : IVehicle
    {
        protected Vehicle(string name)
        {
            Name = name;
        }
        public string Name { get; set; }
        public ISpecs Specs { get; set; }
    }

    public class Fleet
    {
        public IList<IProduct> Vehicles { get; set; }
    }

    public class Specs : ISpecs
    {
        public int Doors { get; set; }
    }
    class Program
    {

        static void Main(string[] args)
        {
            IProduct truck1 = new Truck("The Name")
            {
                Specs = new Specs() { Doors = 2 }
            };

            var fleet = new Fleet()
            {
                Vehicles = new List<IProduct>() { truck1 },
            };

            string serialized1 = JsonSerializer.Serialize(fleet);
            string serialized2 = JsonSerializer.Serialize(fleet, typeof(Fleet));
            string serialized3 = JsonSerializer.Serialize(fleet, fleet.GetType());
        }
    }
}
`

**Comment by bartonjs** [comment]
I neglected to notice that my first bullet the generic specifier got eaten as a bad HTML tag, it should have been `JsonSerializer.Serialize<object>(value)`.

If you want "deeply polymorphic" you need to serialize as object. Any other type, including getting the type by generic inference, is intentionally static typed, not polymorphic.

**Comment by mauricio-bv** [comment]
Do you mean 

`string serialized4 = JsonSerializer.Serialize<Fleet>(fleet);`

I did that and get the same result

```
{
  "Vehicles":
  [
    {"Name":"The Name"}
  ]
}
```

**Comment by bartonjs** [comment]
@mauricio-bv No, literally the word "object", as in `System.Object`.

**Comment by mauricio-bv** [comment]
Oh. I see. Still, I tried that with the same result

`string serialized4 = JsonSerializer.Serialize<object>(fleet);`

Still getting the same result (no Specs property)

`{"Vehicles":[{"Name":"The Name"}]}`

**Comment by mauricio-bv** [comment]
Anyone that can help on my last message?

**Comment by bailei1987** [comment]
I have the same problem.the inherit property can not be transfered to web front.

        [HttpGet("TT")]
        public object TT()
        {
            A obj = new A { Name = "xb" };
            obj.Items.Add(new B { Age = 5 });
            obj.Items.Add(new C { Age = 6, CAge = 7 });
            return obj;
        }
        public class A
        {
            public string Name { get; set; }
            public List<B> Items { get; set; } = new List<B>();
        }
        public class B
        {
            public int Age { get; set; }
        }
        public class C : B
        {
            public int CAge { get; set; }
        }

//api request result expect
{"name":"xb","items":[{"age":5},{"age":6,"cAge":7}]}
//actual
{"name":"xb","items":[{"age":5},{"age":6}]}

**Comment by Arash-Sabet** [comment]
I think it makes perfect sense to have the behavior of the serializer changed especially by considering the issue that @bailei1987 reported. How would serializer behave in signalR context when a polymorphic object is expected? 

Basically a simple JsonSerializer.Serialize(value, ...) should serialize an object properly without having to be worried about its polymorphism.

/cc @bartonjs @terrajobst

**Comment by bartonjs** [comment]
> Basically a simple JsonSerializer.Serialize(value, ...) should serialize an object properly without having to be worried about its polymorphism.

I agree with your statement, but I think I have to fundamentally disagree with your implied conclusion.  Statically typed serialization is more secure (you can understand what data is written ahead of time, so no "I didn't remember this object was serialized and this property shouldn't have been" information disclosure) and able to be pregenerated for much more optimal code.

I'm willing to concede there should be an option to turn it on, but it should neither be the only behavior, nor the default.

**Comment by mauricio-bv** [comment]
> I think it makes perfect sense to have the behavior of the serializer changed especially by considering the issue that @bailei1987 reported. How would serializer behave in signalR context when a polymorphic object is expected?
> Basically a simple JsonSerializer.Serialize(value, ...) should serialize an object properly without having to be worried about its polymorphism.
> /cc @bartonjs @terrajobst

Serialization goes well beyond sending data through the web and a good library should not limit polymorphism/SOLID principles, specially in .Net 3.0 that is a multiplatform framework. I am trying to move to the .Net core 3.0 serializer from Newtonsoft.JSon (which serializes perfectly deep polymorphic cobjects) because it is the recommended serializer for .Net Core 3.0, but I can't compromise my SOLID pattern/Interfaces Segregation.

**Comment by Arash-Sabet** [comment]
> I'm willing to concede there should be an option to turn it on, but it should neither be the only behavior, nor the default.

I think this would be acceptable too. Bottom line something that can turn on and automatically serialize a derived class's object.
